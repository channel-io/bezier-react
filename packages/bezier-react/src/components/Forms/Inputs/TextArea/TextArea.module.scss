@use "../../../../styles/mixins/typography";

.TextArea {
  @include typography.size(14);

  resize: none;

  display: flex;
  align-items: center;

  box-sizing: border-box;
  width: 100%;
  margin: 0;
  padding: 8px 12px;

  color: var(--txt-black-darkest);
  overflow-wrap: anywhere;

  background-color: var(--bg-grey-lightest);
  border: none;
  border-radius: var(--radius-8);
  outline: none;
  box-shadow: var(--input-box-shadow);

  transition: border-color var(--transition-s), box-shadow var(--transition-s);

  &::placeholder {
    color: var(--txt-black-dark);
  }

  &::-webkit-scrollbar {
    display: none;
  }

  &:disabled {
    cursor: not-allowed;
    opacity: var(--opacity-disabled);
  }

  &:focus {
    background-color: var(--bg-white-normal);
    box-shadow: var(--input-box-shadow-focused);
  }

  &[aria-invalid='true'] {
    background-color: var(--bg-white-normal);
    box-shadow: var(--input-box-shadow-invalid);
  }

  &:read-only {
    color: var(--txt-black-darker);
    background-color: var(--bg-grey-lighter);
  }
}
