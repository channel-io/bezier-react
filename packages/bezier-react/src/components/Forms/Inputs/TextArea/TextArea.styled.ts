import TextareaAutosize from 'react-textarea-autosize'

import {
  type SemanticNames,
  Typography,
  hideScrollbars,
  styled,
} from '~/src/foundation'

import { type InterpolationProps } from '~/src/types/Foundation'

import {
  erroredInputWrapperStyle,
  focusedInputWrapperStyle,
  inputPlaceholderStyle,
  inputTextStyle,
  inputWrapperStyle,
} from '~/src/components/Forms/Inputs/mixins'

interface WrapperProps extends InterpolationProps {
  focused: boolean
  hasError?: boolean
  bgColor: SemanticNames
  disabled: boolean
}

const Wrapper = styled.div<WrapperProps>`
  ${inputWrapperStyle};

  box-sizing: border-box;
  display: flex;
  align-items: center;
  background-color: ${({ foundation, bgColor }) => foundation?.theme?.[bgColor]};
  opacity: ${({ disabled }) => disabled && 'var(--opacity-disabled)'};

  ${({ focused }) => focused && focusedInputWrapperStyle}

  ${({ hasError }) => hasError && erroredInputWrapperStyle}

  ${({ foundation }) => foundation?.rounding.round8}

  ${({ foundation }) => foundation?.transition?.getTransitionsCSS(['border-color', 'box-shadow'])};

  ${({ interpolation }) => interpolation}
`

interface TextAreaAutoSizeBaseProps extends InterpolationProps {
  disabled: boolean
  readOnly: boolean
}

const TextAreaAutoSizeBase = styled(TextareaAutosize)<TextAreaAutoSizeBaseProps>`
  box-sizing: border-box;
  width: 100%;
  padding: 8px 12px;
  margin: 0;
  cursor: ${({ disabled }) => (disabled ? 'not-allowed' : 'auto')};
  resize: none;
  background: none;
  border: none;
  outline: none;

  ${hideScrollbars()}

  ${Typography.Size14}

  ${inputTextStyle}

  ${inputPlaceholderStyle}

  &::placeholder {
    ${Typography.Size14}
  }

  ${({ interpolation }) => interpolation}
`

export default {
  Wrapper,
  TextAreaAutoSizeBase,
}
