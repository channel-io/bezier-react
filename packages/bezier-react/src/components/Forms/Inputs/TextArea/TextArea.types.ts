/* External dependencies */
import React from 'react'

/* Internal dependencies */
import type { BezierComponentProps, AdditionalStylableProps } from 'Types/ComponentProps'
import { FormComponentProps } from 'Components/Forms/Form.types'

export enum TextAreaHeight {
  Row3 = 3,
  Row6 = 6,
  Row10 = 10,
  Row16 = 16,
  Row24 = 24,
  Row36 = 36,
}

type TextAreaChangeEventHandler = React.ChangeEventHandler<HTMLTextAreaElement>

interface TextAreaOptions {
  minRows?: TextAreaHeight
  maxRows?: TextAreaHeight
  autoFocus?: boolean
  onFocus?: TextAreaChangeEventHandler
  onBlur?: TextAreaChangeEventHandler
  onChange?: TextAreaChangeEventHandler
}

type OmittedTextareaHTMLAttributes = keyof FormComponentProps | 'onFocus' | 'onBlur'

export default interface TextAreaProps extends
  BezierComponentProps,
  FormComponentProps,
  Omit<React.TextareaHTMLAttributes<HTMLTextAreaElement>, OmittedTextareaHTMLAttributes>,
  AdditionalStylableProps<'wrapper'>,
  TextAreaOptions {}
