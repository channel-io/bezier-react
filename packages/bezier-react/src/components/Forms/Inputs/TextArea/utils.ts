import { type SemanticNames } from '~/src/foundation'

interface CalculateBgColorProps {
  focused: boolean
  hasError: boolean
  readOnly: boolean
}

export function getTextAreaBgColorSemanticName({
  focused,
  hasError,
  readOnly,
}: CalculateBgColorProps): SemanticNames {
  if (readOnly) { return 'bg-grey-lighter' }
  if (focused || hasError) { return 'bg-white-normal' }
  return 'bg-grey-lightest'
}
