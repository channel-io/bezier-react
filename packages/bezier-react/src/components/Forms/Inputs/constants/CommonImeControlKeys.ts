export const COMMON_IME_CONTROL_KEYS = ['Enter', 'Escape', 'Tab', ' ', 'ArrowUp', 'ArrowDown', 'ArrowLeft', 'ArrowRight']
