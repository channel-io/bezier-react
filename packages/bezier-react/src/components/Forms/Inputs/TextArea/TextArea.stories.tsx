/* External dependencies */
import React, { useCallback, useState } from 'react'
import base from 'paths.macro'
import { Story, Meta } from '@storybook/react'

/* Internal dependencies */
import { getObjectFromEnum, getTitle } from 'Utils/storyUtils'
import TextArea from './TextArea'
import TextAreaProps, { TextAreaHeight } from './TextArea.types'

export default {
  title: getTitle(base),
  component: TextArea,
  argTypes: {
    minRows: {
      control: {
        type: 'radio',
        options: getObjectFromEnum(TextAreaHeight),
      },
    },
    maxRows: {
      control: {
        type: 'radio',
        options: getObjectFromEnum(TextAreaHeight),
      },
    },
  },
} as Meta

const Template: Story<TextAreaProps> = (args) => {
  const [value, setValue] = useState('12345')

  const handleChange = useCallback((e: React.ChangeEvent<HTMLTextAreaElement>) => {
    // eslint-disable-next-line no-console
    console.log(e)
    setValue(e.target.value)
  }, [])

  return (
    <div style={{ width: 500 }}>
      <TextArea
        value={value}
        onChange={handleChange}
        {...args}
      />
    </div>
  )
}

export const Primary = Template.bind({})

Primary.args = {
  autoFocus: true,
  readOnly: false,
  disabled: false,
  hasError: false,
  minRows: TextAreaHeight.Row6,
  maxRows: TextAreaHeight.Row10,
  placeholder: 'say hi to autoResizable textarea!',
}
