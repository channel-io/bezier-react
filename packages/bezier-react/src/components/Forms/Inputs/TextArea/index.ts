/* Internal dependencies */
import TextAreaProps, { TextAreaHeight } from './TextArea.types'
import TextArea from './TextArea'

export type {
  TextAreaProps,
}

export {
  TextArea,
  TextAreaHeight,
}
