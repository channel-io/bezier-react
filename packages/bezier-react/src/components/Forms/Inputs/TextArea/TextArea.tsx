/* External dependencies */
import React, { forwardRef, Ref, useRef, useCallback, useState, useLayoutEffect, useMemo } from 'react'

/* Internal dependencies */
import useMergeRefs from 'Hooks/useMergeRefs'
import useFormFieldProps from 'Components/Forms/useFormFieldProps'
import Styled from './TextArea.styled'
import { getTextAreaBgColorSemanticName } from './utils'
import TextAreaProps, { TextAreaHeight } from './TextArea.types'

export const TEXT_AREA_TEST_ID = 'bezier-react-text-area'

function TextArea({
  interpolation,
  wrapperInterpolation,
  className,
  wrapperClassName,
  style,
  wrapperStyle,
  testId = TEXT_AREA_TEST_ID,
  minRows = TextAreaHeight.Row6,
  maxRows = TextAreaHeight.Row6,
  autoFocus = false,
  value = '',
  onFocus,
  onBlur,
  onChange,
  ...rest
}: TextAreaProps,
forwardedRef: Ref<HTMLTextAreaElement>,
) {
  const {
    disabled,
    readOnly,
    hasError,
    ...ownProps
  } = useFormFieldProps(rest)

  const inputRef = useRef<HTMLTextAreaElement | null>(null)
  const mergedInputRef = useMergeRefs<HTMLTextAreaElement>(inputRef, forwardedRef)

  const activeInput = !disabled && !readOnly

  const [focused, setFocused] = useState(false)

  const bgColorSemanticName = useMemo(() => (
    getTextAreaBgColorSemanticName({
      focused,
      hasError,
      readOnly,
    })
  ), [
    focused,
    hasError,
    readOnly,
  ])

  const handleFocus = useCallback((event: React.FocusEvent<HTMLTextAreaElement>) => {
    if (!activeInput) { return }
    setFocused(true)
    onFocus?.(event)
  }, [
    onFocus,
    activeInput,
  ])

  const handleBlur = useCallback((event: React.FocusEvent<HTMLTextAreaElement>) => {
    setFocused(false)
    onBlur?.(event)
  }, [onBlur])

  // eslint-disable-next-line prefer-arrow-callback
  useLayoutEffect(function initialAutoFocus() {
    function setSelectionToEnd() {
      inputRef.current?.setSelectionRange(inputRef.current?.value.length, inputRef.current?.value.length)
    }

    if (autoFocus) {
      inputRef.current?.focus()
      setSelectionToEnd()
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  return (
    <Styled.Wrapper
      className={className}
      interpolation={interpolation}
      focused={focused}
      hasError={hasError}
      disabled={disabled}
      bgColor={bgColorSemanticName}
      data-testid={testId}
    >
      <Styled.TextAreaAutoSizeBase
        {...ownProps}
        ref={mergedInputRef}
        value={value}
        disabled={disabled}
        readOnly={readOnly}
        maxRows={maxRows}
        minRows={minRows}
        onChange={onChange}
        onFocus={handleFocus}
        onBlur={handleBlur}
      />
    </Styled.Wrapper>
  )
}

export default forwardRef(TextArea)
