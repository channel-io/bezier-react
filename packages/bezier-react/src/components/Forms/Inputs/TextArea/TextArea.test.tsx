import React from 'react'

import { fireEvent } from '@testing-library/dom'
import { act } from '@testing-library/react'

import { LightFoundation } from '~/src/foundation'

import disabledOpacity from '~/src/constants/DisabledOpacity'
import { render } from '~/src/utils/test'

import { COMMON_IME_CONTROL_KEYS } from '~/src/components/Forms/Inputs/constants/CommonImeControlKeys'

import TextArea, { TEXT_AREA_TEST_ID } from './TextArea'
import type TextAreaProps from './TextArea.types'
import { getTextAreaBgColorSemanticName } from './utils'

// SEE ALSO: https://github.com/Andarist/react-textarea-autosize#how-to-test-it-with-jest-and-react-test-renderer-if-you-need-ref
describe('TextArea 테스트 >', () => {
  let props: TextAreaProps

  beforeEach(() => {
    jest.useFakeTimers()
    props = {}
  })

  const renderComponent = (optionProps?: TextAreaProps) => render(
    <TextArea {...props} {...optionProps} />,
  )

  it('SnapShot 테스트 >', () => {
    const { container } = renderComponent()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('TextArea는 기본 attribute들을 가져야 한다', () => {
    const { getByTestId } = renderComponent()
    const rendered = getByTestId(TEXT_AREA_TEST_ID)
    const textareaElement = rendered.getElementsByTagName('textarea')[0]

    expect(textareaElement).not.toHaveAttribute('readOnly')
    expect(textareaElement).not.toHaveAttribute('disabled')
    expect(textareaElement).not.toHaveAttribute('placeholder')
    expect(textareaElement).not.toHaveAttribute('maxRows')
    expect(textareaElement).not.toHaveAttribute('minRows')
  })

  it('placeholder가 주입되었을 때 주입되는 값과 동일한 "placeholder"를 가져야 한다', () => {
    const PLACEHOLDER_TEXT = 'this is placeholder'
    const { getByTestId } = renderComponent({ placeholder: PLACEHOLDER_TEXT })
    const rendered = getByTestId(TEXT_AREA_TEST_ID)
    const textareaElement = rendered.getElementsByTagName('textarea')[0]

    expect(textareaElement).toHaveAttribute('placeholder', PLACEHOLDER_TEXT)
  })

  it('disabled 나 readOnly prop이 주입되었을 때 주입되는 attribute를 가져야 한다', () => {
    const { getByTestId } = renderComponent({ disabled: true, readOnly: true })
    const rendered = getByTestId(TEXT_AREA_TEST_ID)
    const textareaElement = rendered.getElementsByTagName('textarea')[0]
    expect(textareaElement).toHaveAttribute('readOnly')
    expect(textareaElement).toHaveAttribute('disabled')
  })

  it('disabled prop이 주입되었을 때는 root wrapper의 opacity가 0.4이어야 한다', () => {
    const { getByTestId } = renderComponent({ disabled: true })
    const rendered = getByTestId(TEXT_AREA_TEST_ID)
    expect(rendered).toHaveStyle(`opacity: ${disabledOpacity}`)
  })

  it('focus 상태일 때는 그에 맞는 shadow 스타일을 가져야 한다', () => {
    const onFocus = jest.fn()
    const { getByTestId } = renderComponent({ onFocus })
    const rendered = getByTestId(TEXT_AREA_TEST_ID)
    const textareaElement = rendered.getElementsByTagName('textarea')[0]
    act(() => {
      textareaElement.focus()
    })

    jest.advanceTimersByTime(1000)
    // eslint-disable-next-line max-len
    expect(rendered).toHaveStyle(`box-shadow: 0 0 0 3px ${LightFoundation.theme['bgtxt-blue-light']}, inset 0 0 0 1px ${LightFoundation.theme['bgtxt-blue-normal']}`)
  })

  it('error 상태일 때는 그에 맞는 shadow 스타일을 가져야 한다', () => {
    const onFocus = jest.fn()
    const { getByTestId } = renderComponent({ onFocus, hasError: true })
    const rendered = getByTestId(TEXT_AREA_TEST_ID)

    jest.advanceTimersByTime(1000)
    // eslint-disable-next-line max-len
    expect(rendered).toHaveStyle(`box-shadow: 0 0 0 3px ${LightFoundation.theme['bgtxt-orange-light']}, inset 0 0 0 1px ${LightFoundation.theme['bgtxt-orange-normal']}`)
  })

  describe('onFocus 테스트 >', () => {
    it('readOnly가 주입됐다면 onFocus 가 안 불려야 한다', () => {
      const onFocus = jest.fn()
      const { getByTestId } = renderComponent({ onFocus, readOnly: true })
      const rendered = getByTestId(TEXT_AREA_TEST_ID)
      const textareaElement = rendered.getElementsByTagName('textarea')[0]
      act(() => {
        textareaElement.focus()
      })

      expect(onFocus).not.toBeCalled()
    })

    it('disabled가 주입됐다면 onFocus 가 안 불려야 한다', () => {
      const onFocus = jest.fn()
      const { getByTestId } = renderComponent({ onFocus, disabled: true })
      const rendered = getByTestId(TEXT_AREA_TEST_ID)
      const textareaElement = rendered.getElementsByTagName('textarea')[0]
      act(() => {
        textareaElement.focus()
      })

      expect(onFocus).not.toBeCalled()
    })

    it('disabled, readOnly가 주입됐다면 onFocus 가 안 불려야 한다', () => {
      const onFocus = jest.fn()
      const { getByTestId } = renderComponent({ onFocus, disabled: true, readOnly: true })
      const rendered = getByTestId(TEXT_AREA_TEST_ID)
      const textareaElement = rendered.getElementsByTagName('textarea')[0]
      act(() => {
        textareaElement.focus()
      })

      expect(onFocus).not.toBeCalled()
    })

    it('readOnly, disabled 가 주입되지 않았다면 onFocus는 불려야 한다', () => {
      const onFocus = jest.fn()
      const { getByTestId } = renderComponent({ onFocus })
      const rendered = getByTestId(TEXT_AREA_TEST_ID)
      const textareaElement = rendered.getElementsByTagName('textarea')[0]
      act(() => {
        textareaElement.focus()
      })

      expect(onFocus).toBeCalled()
    })
  })

  describe('onChange 테스트 >', () => {
    it('정상적인 상황에서 잘 불린다', () => {
      const onChange = jest.fn()
      const { getByTestId } = renderComponent({ onChange })
      const rendered = getByTestId(TEXT_AREA_TEST_ID)
      const textareaElement = rendered.getElementsByTagName('textarea')[0]

      act(() => {
        fireEvent.change(textareaElement, { target: { value: 'test' } })
      })

      expect(onChange).toBeCalled()
    })
  })

  describe('onBlur 테스트 >', () => {
    it('정상적인 상황에서 잘 불린다', () => {
      const onBlur = jest.fn()
      const { getByTestId } = renderComponent({ onBlur })
      const rendered = getByTestId(TEXT_AREA_TEST_ID)
      const textareaElement = rendered.getElementsByTagName('textarea')[0]
      act(() => {
        textareaElement.focus()
        textareaElement.blur()
      })
      expect(onBlur).toBeCalled()
    })
  })

  describe('autoFocus 테스트 >', () => {
    it('autoFocus를 주입하면 focus 상태로 되어야 한다', () => {
      const { getByTestId } = renderComponent({ autoFocus: true })
      const rendered = getByTestId(TEXT_AREA_TEST_ID)
      const textareaElement = rendered.getElementsByTagName('textarea')[0]

      expect(textareaElement).toEqual(document.activeElement)
    })

    it('selection 이 가장 끝에 위치하여야 한다', () => {
      const TEST_INITIAL_VALUE = 'test value'
      const { getByTestId } = renderComponent({ autoFocus: true, value: TEST_INITIAL_VALUE })
      const rendered = getByTestId(TEXT_AREA_TEST_ID)
      const textareaElement = rendered.getElementsByTagName('textarea')[0]
      expect(textareaElement.selectionEnd).toEqual(TEST_INITIAL_VALUE.length)
    })
  })

  describe('Keyboard event handlers for common ime control keys should not be called while composing ', () => {
    it('onKeyDown', () => {
      const onKeyDown = jest.fn()
      const { getByTestId } = renderComponent({ onKeyDown })
      const rendered = getByTestId(TEXT_AREA_TEST_ID)
      const textareaElement = rendered.getElementsByTagName('textarea')[0]

      COMMON_IME_CONTROL_KEYS.forEach((key) => {
        const isCompositionStartFired = fireEvent.compositionStart(textareaElement)
        fireEvent.keyDown(textareaElement, { key, isComposing: isCompositionStartFired })
        expect(onKeyDown).not.toBeCalled()
      })
    })

    it('onKeyUp', () => {
      const onKeyUp = jest.fn()
      const { getByTestId } = renderComponent({ onKeyUp })
      const rendered = getByTestId(TEXT_AREA_TEST_ID)
      const textareaElement = rendered.getElementsByTagName('textarea')[0]

      COMMON_IME_CONTROL_KEYS.forEach((key) => {
        const isCompositionStartFired = fireEvent.compositionStart(textareaElement)
        fireEvent.keyUp(textareaElement, { key, isComposing: isCompositionStartFired })
        expect(onKeyUp).not.toBeCalled()
      })
    })
  })
})

describe('TextArea util test >', () => {
  describe('getTextAreaBgColorSemanticNames 테스트', () => {
    it('readOnly 이면 다른 값이 무엇이든 bg-grey-lighter를 반환한다.', () => {
      const result1 = getTextAreaBgColorSemanticName({
        focused: true,
        hasError: true,
        readOnly: true,
      })

      const result2 = getTextAreaBgColorSemanticName({
        focused: false,
        hasError: false,
        readOnly: true,
      })

      expect(result1).toBe('bg-grey-lighter')
      expect(result2).toBe('bg-grey-lighter')
    })

    it('!readOnly이고, focused이거나 hasError이면 bg-white-normal을 반환한다', () => {
      const result1 = getTextAreaBgColorSemanticName({
        focused: true,
        hasError: false,
        readOnly: false,
      })

      const result2 = getTextAreaBgColorSemanticName({
        focused: false,
        hasError: true,
        readOnly: false,
      })

      expect(result1).toBe('bg-white-normal')
      expect(result2).toBe('bg-white-normal')
    })

    it('모두 아니면 bg-grey-lightest를 반환한다', () => {
      const result1 = getTextAreaBgColorSemanticName({
        focused: false,
        hasError: false,
        readOnly: false,
      })

      expect(result1).toBe('bg-grey-lightest')
    })
  })
})
