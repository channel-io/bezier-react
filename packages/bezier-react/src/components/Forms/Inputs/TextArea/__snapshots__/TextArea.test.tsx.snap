// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea 테스트 > SnapShot 테스트 > 1`] = `
.c0 {
  box-shadow: 0 1px 2px #0000000D, inset 0 0 0 1px #00000026;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #FCFCFC;
  overflow: hidden;
  border-radius: 8px;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: border-color,box-shadow;
  transition-property: border-color,box-shadow;
}

.c1 {
  box-sizing: border-box;
  width: 100%;
  padding: 8px 12px;
  margin: 0;
  cursor: auto;
  resize: none;
  background: none;
  border: none;
  outline: none;
  -ms-overflow-style: none;
  font-size: 1.4rem;
  line-height: 1.8rem;
  color: #000000D9;
}

.c1::-webkit-scrollbar {
  display: none;
}

.c1::-webkit-input-placeholder {
  color: #00000066;
}

.c1::-moz-placeholder {
  color: #00000066;
}

.c1:-ms-input-placeholder {
  color: #00000066;
}

.c1::placeholder {
  color: #00000066;
}

.c1::-webkit-input-placeholder {
  font-size: 1.4rem;
  line-height: 1.8rem;
}

.c1::-moz-placeholder {
  font-size: 1.4rem;
  line-height: 1.8rem;
}

.c1:-ms-input-placeholder {
  font-size: 1.4rem;
  line-height: 1.8rem;
}

.c1::placeholder {
  font-size: 1.4rem;
  line-height: 1.8rem;
}

<div
  class="c0"
  data-testid="bezier-react-text-area"
>
  <textarea
    class="c1"
    foundation="[object Object]"
  />
</div>
`;
